common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+    this._errorHandler = (err: Error): void => {
+      this._fsm.handle('detach', undefined, err);
+    };
+
+    this._amqp10Client = amqp10Client;
+    /*Codes_SRS_NODE_AMQP_CBS_16_001: [The `constructor` shall instanciate a `SenderLink` object for the `$cbs` endpoint using a custom policy `{encoder: function(body) { return body;}}` which forces the amqp layer to send the token as an amqp value in the body.]*/
+    this._senderLink = new SenderLink(ClaimsBasedSecurityAgent._putTokenSendingEndpoint, { encoder: (body) => body }, this._amqp10Client);
+    this._senderLink.on('error', this._errorHandler);
+
+    /*Codes_SRS_NODE_AMQP_CBS_16_002: [The `constructor` shall instanciate a `ReceiverLink` object for the `$cbs` endpoint.]*/
+    this._receiverLink = new ReceiverLink(ClaimsBasedSecurityAgent._putTokenReceivingEndpoint, null, this._amqp10Client);
+    this._receiverLink.on('error', this._errorHandler);
+
+    this._putTokenQueue = [];
+
+    this._fsm = new machina.Fsm({
+      initialState: 'detached',
+      states: {
+        detached: {
+          _onEnter: (callback, err) => {
+            let tokenOperation = this._putTokenQueue.shift();
+            while (tokenOperation) {
+              tokenOperation.callback(err);
+              tokenOperation = this._putTokenQueue.shift();
+            }
+
+            /*Codes_SRS_NODE_AMQP_CBS_16_006: [If given as an argument, the `attach` method shall call `callback` with a standard `Error` object if any link fails to attach.]*/
+            if (callback) {
+              callback(err);
+            }
+          },
+          attach: (callback) => {
+            this._fsm.transition('attaching', callback);
+          },
+          detach: (callback) => { if (callback) callback(); },
+          forceDetach: () => { return; },
+          putToken: (audience, token, callback) => {
+            this._putTokenQueue.push({
+              audience: audience,
+              token: token,
+              callback: callback
+            });
+            this._fsm.transition('attaching');
+          }
+         },
+        attaching: {
+          _onEnter: (callback) => {
+            this._senderLink.attach((err) => {
+              if (err) {
+                this._fsm.transition('detached', callback, err);
+              } else {
+                this._receiverLink.attach((err) => {
+                  if (err) {
+                    this._fsm.transition('detaching', callback, err);
+                  } else {
+                    this._receiverLink.on('message', (msg) => {
+                      //
+                      // Regardless of whether we found the put token in the list of outstanding
+                      // operations, accept it.  This could be a put token that we previously
+                      // timed out.  Be happy.  It made it home, just too late to be useful.
+                      //
+                      /*Codes_SRS_NODE_AMQP_CBS_16_020: [All responses shall be completed.]*/
```

Comment:
>completed [](start = 80, length = 9)

Shouldn't this say "all responses should be accepted"?  (I don't remember if AMQP has "complete" and "accept" as 2 different dispositions, but using the right word here might be significant)

Reply:
"Accept" is the right word and you are right. i've been trying to remove our use of "Complete" everywhere

---
In reply to: [139538944](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139538944) [](ancestors = 139538944)